### ISSUE TITLE
Background invalidation of sliding windows

### ISSUE THREAD

**Issue by dev1ant** [pipelinedb/pipelinedb#1130]
```
pgload01g/test M # select pipeline_version();
                                                                          pipeline_version
--------------------------------------------------------------------------------------------------------------------------------------------------------------------
 PipelineDB 0.8.0 at revision bc8c4e3decbcc219689330987ea46ddb6b3edd34 on x86_64-unknown-linux-gnu, compiled by gcc (GCC) 4.4.7 20120313 (Red Hat 4.4.7-11), 64-bit
(1 row)

Time: 0.298 ms
pgload01g/test M # \dS+ recent_sessions
                       View "public.recent_sessions"
   Column   |           Type           | Modifiers | Storage  | Description
------------+--------------------------+-----------+----------+-------------
 uid        | bigint                   |           | plain    |
 session_id | text                     |           | extended |
 last_ts    | timestamp with time zone |           | plain    |
View definition:
 SELECT uid,
    session_id,
    max(ts) AS last_ts
   FROM ONLY stream1
  WHERE arrival_timestamp > (clock_timestamp() - '00:15:00'::interval) AND session_id IS NOT NULL
  GROUP BY uid, session_id;

pgload01g/test M # \dS+ recent_sessions_mrel0
                           Table "public.recent_sessions_mrel0"
   Column   |           Type           | Modifiers | Storage  | Stats target | Description
------------+--------------------------+-----------+----------+--------------+-------------
 _0         | timestamp with time zone |           | plain    |              |
 uid        | bigint                   |           | plain    |              |
 session_id | text                     |           | extended |              |
 last_ts    | timestamp with time zone |           | plain    |              |
Indexes:
    "recent_sessions_mrel0_expr_idx" btree (ls_hash_group(_0, uid, session_id))
    "recent_sessions_mrel0_uid_idx" btree (uid)
Options: fillfactor=50

pgload01g/test M # select count(*) from recent_sessions;
 count
-------
     0
(1 row)

Time: 1228.627 ms
pgload01g/test M # select count(*) from recent_sessions_mrel0 ;
  count
---------
 5000008
(1 row)

Time: 719.131 ms
pgload01g/test M # select clock_timestamp(), min(last_ts), clock_timestamp() - min(last_ts) as diff from recent_sessions_mrel0 ;
        clock_timestamp        |             min              |      diff
-------------------------------+------------------------------+-----------------
 2015-08-24 11:44:31.624587+03 | 2015-08-24 10:29:28.20038+03 | 01:15:03.424211
(1 row)

Time: 914.546 ms
pgload01g/test M #
```

Am I right that these tuples should have been removed an hour ago? Is there a way to trigger such process manually?

**Comment by usmanm** [comment]
Is the auto-vacuumer enabled and running?

**Comment by dev1ant** [comment]
Yep.

```
root@pgload01g ~ # psql test
psql (9.4.4)
Type "help" for help.

pgload01g/test M # select name, setting from pg_settings where name ilike 'autovacuum%';
                name                 |  setting
-------------------------------------+-----------
 autovacuum                          | on
 autovacuum_analyze_scale_factor     | 0.01
 autovacuum_analyze_threshold        | 50
 autovacuum_freeze_max_age           | 200000000
 autovacuum_max_workers              | 10
 autovacuum_multixact_freeze_max_age | 400000000
 autovacuum_naptime                  | 15
 autovacuum_vacuum_cost_delay        | 5
 autovacuum_vacuum_cost_limit        | 2000
 autovacuum_vacuum_scale_factor      | 0.001
 autovacuum_vacuum_threshold         | 200
 autovacuum_work_mem                 | -1
(12 rows)

Time: 6.031 ms
pgload01g/test M # \q
root@pgload01g ~ # ps aux | grep autovacuum
postgres  1911 77.5  0.0 1203780 39184 ?       Rs   19:09   1:05 pipeline: autovacuum worker process
postgres  2917  102  0.0 153980 11204 ?        Rs   19:10   0:09 pipeline: autovacuum worker process
root      2977  0.0  0.0 103248   988 pts/1    S+   19:10   0:00 grep autovacuum
postgres 31053  0.0  0.1 401096 249288 ?       Ss   Aug23   0:03 pipeline: autovacuum launcher process
root@pgload01g ~ # psql test
psql (9.4.4)
Type "help" for help.

pgload01g/test M # select * from pg_stat_activity where pid in (1911, 2917);
-[ RECORD 1 ]----+--------------------------------------------------------
datid            | 16384
datname          | test
pid              | 2917
usesysid         | 10
usename          | postgres
application_name |
client_addr      | [null]
client_hostname  | [null]
client_port      | [null]
backend_start    | 2015-08-24 19:10:17.783326+03
xact_start       | 2015-08-24 19:10:35.180781+03
query_start      | 2015-08-24 19:10:35.180781+03
state_change     | 2015-08-24 19:10:35.180782+03
waiting          | f
state            | active
backend_xid      | 614062
backend_xmin     | 614062
query            | autovacuum: VACUUM ANALYZE public.recent_sessions_mrel0
-[ RECORD 2 ]----+--------------------------------------------------------
datid            | 16384
datname          | test
pid              | 1911
usesysid         | 10
usename          | postgres
application_name |
client_addr      | [null]
client_hostname  | [null]
client_port      | [null]
backend_start    | 2015-08-24 19:09:02.658111+03
xact_start       | 2015-08-24 19:09:19.132584+03
query_start      | 2015-08-24 19:09:19.132584+03
state_change     | 2015-08-24 19:09:19.132585+03
waiting          | f
state            | active
backend_xid      | [null]
backend_xmin     | 612618
query            | autovacuum: VACUUM public.user_activity_mrel0

Time: 2.789 ms
pgload01g/test M #
```

**Comment by usmanm** [comment]
Seems like `public.recent_sessions_mrel0` is being vacuumed. Once the vacuuming is done, can you check if the discarded rows have been removed?

**Comment by dev1ant** [comment]
These vacuums are now triggered with a cron job that does deletes manually:
```
*/1 * * * *    postgres flock -n -w 10 /tmp/remove_recent_sessions.lock psql test -c "delete from recent_sessions_mrel0 where _0 < clock_timestamp() - '15 minutes'::interval;" >/dev/null 2>&1
```

Seems, that the problem is that until you do some updates/deletes manually, vacuum would not be started because for such view you can have only inserts (as in my case)... And it doesn't matter how aggressive your autovacuum settings are.

It doesn't seem to be the pipelinedb problem, so the issue could be closed, I think. It would be nice to have a solution for that, but I don't see a good way to do it...

**Comment by derekjn** [comment]
Technically a vacuum should still be triggered because of this:

https://github.com/pipelinedb/pipelinedb/blob/master/src/backend/postmaster/autovacuum.c#L2786

**Comment by usmanm** [comment]
Moreover, inserts into streams will trigger updates of continuous views if the row being affected already exists, so dead tuples are constantly being generated.

**Comment by dev1ant** [comment]
> Technically a vacuum should still be triggered because of this:
> https://github.com/pipelinedb/pipelinedb/blob/master/src/backend/postmaster/autovacuum.c#L2786

Hm, maybe I will turn on DEBUG3 log level to get more details.

> inserts into streams will trigger updates of continuous views if the row being affected already exists, so dead tuples are constantly being generated.

Yep, but in my case each row contains unique ID, so there are no dead tuples.

**Comment by usmanm** [comment]
Hey @dev1ant, did you get a chance to check the DEBUG3 logs? Or if there's someway we could simulate your workload to see if we can reproduce this?

**Comment by dev1ant** [comment]
Not yet, sorry. I will do it until the end of the week.

**Comment by usmanm** [comment]
No problem man! Whenever you can. We appreciate your digging into the matter.

**Comment by dev1ant** [comment]
No, it seems that vacuuming does not clear discarded tuples. Here is some diagnostics:
```
pgload01g/test M # select pipeline_version();
                                                         pipeline_version
-----------------------------------------------------------------------------------------------------------------------------------
 PipelineDB 0.8.0 at revision 3013d7d on x86_64-unknown-linux-gnu, compiled by gcc (GCC) 4.4.7 20120313 (Red Hat 4.4.7-11), 64-bit
(1 row)

Time: 0.740 ms
pgload01g/test M # \dS+ recent_sessions;
                       View "public.recent_sessions"
   Column   |           Type           | Modifiers | Storage  | Description
------------+--------------------------+-----------+----------+-------------
 uid        | bigint                   |           | plain    |
 session_id | text                     |           | extended |
 last_ts    | timestamp with time zone |           | plain    |
View definition:
 SELECT uid,
    session_id,
    max(ts) AS last_ts
   FROM ONLY stream1
  WHERE arrival_timestamp > (clock_timestamp() - '00:15:00'::interval) AND session_id IS NOT NULL
  GROUP BY uid, session_id;

pgload01g/test M # select reltoastrelid from pg_class where relname = 'recent_sessions_mrel0';
 reltoastrelid
---------------
         16459
(1 row)

Time: 0.580 ms
pgload01g/test M # \q
postgres@pgload01g ~ $ /usr/pgsql-9.4/bin/oid2name -d test -o 16459
From database "test":
  Filenode      Table Name
--------------------------
     16634  pg_toast_16456
postgres@pgload01g ~ $
```

Here is the way I create the load:
```
postgres@pgload01g ~ $ cat /tmp/insert_uniq.sql
\setrandom some_id 0 9
INSERT INTO stream1 (uid, ts, service, session_id) VALUES(round(random()*200000000), current_timestamp, :some_id::text, md5(random()::text));
postgres@pgload01g ~ $ /usr/pgsql-9.4/bin/pgbench -n -c 4 -j 1 -T 3600 -P 60 -f /tmp/insert_uniq.sql test
```

And the discarded rows are not removed:
```
pgload01g/test M # select current_timestamp, last_ts from recent_sessions_mrel0 where last_ts <= (clock_timestamp() - '00:15:00'::interval) order by last_ts limit 1;
              now              |            last_ts
-------------------------------+-------------------------------
 2015-08-28 12:04:28.307835+03 | 2015-08-28 11:41:50.320975+03
(1 row)

Time: 10088.564 ms
pgload01g/test M # select current_timestamp, last_ts from recent_sessions_mrel0 where last_ts <= (clock_timestamp() - '00:15:00'::interval) order by last_ts limit 1;
              now              |            last_ts
-------------------------------+-------------------------------
 2015-08-28 12:09:04.092828+03 | 2015-08-28 11:41:50.320975+03
(1 row)

Time: 12153.477 ms
pgload01g/test M # vacuum recent_sessions_mrel0 ;
VACUUM
Time: 28440.537 ms
pgload01g/test M # select current_timestamp, last_ts from recent_sessions_mrel0 where last_ts <= (clock_timestamp() - '00:15:00'::interval) order by last_ts limit 1;
              now              |            last_ts
-------------------------------+-------------------------------
 2015-08-28 12:10:40.648517+03 | 2015-08-28 11:41:50.320975+03
(1 row)

Time: 12412.407 ms
pgload01g/test M #
```

In the logs with DEBUG3 log-level I do see the following:
```
< 2015-08-28 12:08:17.414 MSK >DEBUG:  recent_sessions_mrel0: vac: 0 (threshold 3500), anl: 177344 (threshold 33049)
< 2015-08-28 12:08:17.415 MSK >DEBUG:  pg_toast_16456: vac: 0 (threshold 200), anl: 0 (threshold 50)
< 2015-08-28 12:08:26.858 MSK >DEBUG:  recent_sessions_mrel0: vac: 0 (threshold 3500), anl: 68350 (threshold 33049)
< 2015-08-28 12:08:31.756 MSK >DEBUG:  recent_sessions_mrel0: vac: 0 (threshold 200), anl: 0 (threshold 50)
< 2015-08-28 12:08:32.263 MSK >DEBUG:  recent_sessions_mrel0: vac: 0 (threshold 3524), anl: 161393 (threshold 33288)
< 2015-08-28 12:08:32.264 MSK >DEBUG:  pg_toast_16456: vac: 0 (threshold 200), anl: 0 (threshold 50)
< 2015-08-28 12:08:41.803 MSK >DEBUG:  recent_sessions_mrel0: vac: 0 (threshold 3524), anl: 86324 (threshold 33288)
< 2015-08-28 12:08:46.759 MSK >DEBUG:  recent_sessions_mrel0: vac: 0 (threshold 200), anl: 0 (threshold 50)
< 2015-08-28 12:08:47.508 MSK >DEBUG:  recent_sessions_mrel0: vac: 0 (threshold 3549), anl: 180384 (threshold 33540)
< 2015-08-28 12:08:47.508 MSK >DEBUG:  pg_toast_16456: vac: 0 (threshold 200), anl: 0 (threshold 50)
< 2015-08-28 12:08:56.970 MSK >DEBUG:  recent_sessions_mrel0: vac: 0 (threshold 3549), anl: 93404 (threshold 33540)
< 2015-08-28 12:09:01.762 MSK >DEBUG:  recent_sessions_mrel0: vac: 0 (threshold 200), anl: 0 (threshold 50)
< 2015-08-28 12:09:02.265 MSK >DEBUG:  recent_sessions_mrel0: vac: 0 (threshold 3574), anl: 176988 (threshold 33789)
< 2015-08-28 12:09:02.265 MSK >DEBUG:  pg_toast_16456: vac: 0 (threshold 200), anl: 0 (threshold 50)
< 2015-08-28 12:09:11.803 MSK >DEBUG:  recent_sessions_mrel0: vac: 0 (threshold 3574), anl: 80118 (threshold 33789)
< 2015-08-28 12:09:16.763 MSK >DEBUG:  recent_sessions_mrel0: vac: 0 (threshold 200), anl: 0 (threshold 50)
< 2015-08-28 12:09:19.234 MSK >DEBUG:  recent_sessions_mrel0: vac: 0 (threshold 3598), anl: 175014 (threshold 34031)
< 2015-08-28 12:09:19.234 MSK >DEBUG:  pg_toast_16456: vac: 0 (threshold 200), anl: 0 (threshold 50)
< 2015-08-28 12:09:29.323 MSK >DEBUG:  recent_sessions_mrel0: vac: 0 (threshold 3598), anl: 114947 (threshold 34031)
< 2015-08-28 12:09:31.763 MSK >DEBUG:  recent_sessions_mrel0: vac: 0 (threshold 200), anl: 0 (threshold 50)
< 2015-08-28 12:09:32.773 MSK >DEBUG:  recent_sessions_mrel0: vac: 0 (threshold 3598), anl: 173916 (threshold 34031)
```

When I try to run `vacuum` on CV manually I hit the error:
```
pgload01g/test M # vacuum recent_sessions ;
ERROR:  XX000: tuple concurrently updated
LOCATION:  simple_heap_delete, heapam.c:3030
Time: 19816.552 ms
pgload01g/test M #
```
And logs change to the following:
```
< 2015-08-28 12:10:02.177 MSK >DEBUG:  recent_sessions_mrel0: vac: 18082356 (threshold 3663), anl: 18724998 (threshold 34679)
< 2015-08-28 12:10:02.177 MSK >DEBUG:  pg_toast_16456: vac: 0 (threshold 200), anl: 0 (threshold 50)
< 2015-08-28 12:10:11.571 MSK >DEBUG:  recent_sessions_mrel0: vac: 2831007 (threshold 3663), anl: 41196 (threshold 34679)
< 2015-08-28 12:10:16.768 MSK >DEBUG:  recent_sessions_mrel0: vac: 0 (threshold 200), anl: 0 (threshold 50)
< 2015-08-28 12:10:19.424 MSK >DEBUG:  recent_sessions_mrel0: vac: 2831007 (threshold 3663), anl: 143852 (threshold 34679)
< 2015-08-28 12:10:19.424 MSK >DEBUG:  pg_toast_16456: vac: 0 (threshold 200), anl: 0 (threshold 50)
< 2015-08-28 12:10:31.770 MSK >DEBUG:  recent_sessions_mrel0: vac: 0 (threshold 200), anl: 0 (threshold 50)
< 2015-08-28 12:10:32.865 MSK >DEBUG:  recent_sessions_mrel0: vac: 2831007 (threshold 3663), anl: 387103 (threshold 34679)
< 2015-08-28 12:10:32.866 MSK >DEBUG:  pg_toast_16456: vac: 0 (threshold 200), anl: 0 (threshold 50)
< 2015-08-28 12:10:46.771 MSK >DEBUG:  recent_sessions_mrel0: vac: 0 (threshold 200), anl: 0 (threshold 50)
< 2015-08-28 12:10:47.810 MSK >DEBUG:  recent_sessions_mrel0: vac: 2831007 (threshold 3663), anl: 645438 (threshold 34679)
< 2015-08-28 12:10:47.810 MSK >DEBUG:  pg_toast_16456: vac: 0 (threshold 200), anl: 0 (threshold 50)
< 2015-08-28 12:11:01.783 MSK >DEBUG:  recent_sessions_mrel0: vac: 0 (threshold 200), anl: 0 (threshold 50)
< 2015-08-28 12:11:03.504 MSK >DEBUG:  recent_sessions_mrel0: vac: 2831007 (threshold 3663), anl: 893099 (threshold 34679)
< 2015-08-28 12:11:03.505 MSK >DEBUG:  pg_toast_16456: vac: 0 (threshold 200), anl: 0 (threshold 50)
< 2015-08-28 12:11:15.293 MSK >DEBUG:  recent_sessions_mrel0: vac: 2831007 (threshold 3663), anl: 1062867 (threshold 34679)
< 2015-08-28 12:11:16.775 MSK >DEBUG:  recent_sessions_mrel0: vac: 0 (threshold 200), anl: 0 (threshold 50)
< 2015-08-28 12:11:18.923 MSK >DEBUG:  recent_sessions_mrel0: vac: 2831007 (threshold 3663), anl: 1135375 (threshold 34679)
```

But tuples are still not removed from the table:
```
pgload01g/test M # select current_timestamp, last_ts from recent_sessions_mrel0 where last_ts <= (clock_timestamp() - '00:15:00'::interval) order by last_ts limit 1;
              now              |            last_ts
-------------------------------+-------------------------------
 2015-08-28 12:21:15.731373+03 | 2015-08-28 11:41:50.320975+03
(1 row)

Time: 17569.359 ms
pgload01g/test M #
```

**Comment by usmanm** [comment]
I think we have seen the tuple concurrently issue before and thought we'd resolved it. Will dig into it more.

**Comment by derekjn** [comment]
@dev1ant https://github.com/pipelinedb/pipelinedb/commit/fa69b5db40318eb44ee8c4fae0dfd78c28417359 should resolve the issue of tuples not being invalidated in the background. I'm not sure that the concurrent update issue is related, so please let us know if it persists and you're able to reproduce it. Thanks for helping out so much, we appreciate it!

**Comment by dev1ant** [comment]
No, that did not help :(
```
pgload01g/test M # select pipeline_version();
                                                         pipeline_version
-----------------------------------------------------------------------------------------------------------------------------------
 PipelineDB 0.8.0 at revision fa69b5d on x86_64-unknown-linux-gnu, compiled by gcc (GCC) 4.4.7 20120313 (Red Hat 4.4.7-11), 64-bit
(1 row)

Time: 0.566 ms
pgload01g/test M # show autovacuum_vacuum_cost_delay ;
 autovacuum_vacuum_cost_delay
------------------------------
 0
(1 row)

Time: 0.304 ms
pgload01g/test M # select clock_timestamp(), min(last_ts), clock_timestamp()-min(last_ts) as diff from recent_sessions_mrel0 ;
        clock_timestamp        |              min              |      diff
-------------------------------+-------------------------------+-----------------
 2015-09-04 11:22:48.434444+03 | 2015-09-04 10:56:51.902241+03 | 00:25:56.532205
(1 row)

Time: 4585.011 ms
pgload01g/test M # \dS+ recent_sessions
                       View "public.recent_sessions"
   Column   |           Type           | Modifiers | Storage  | Description
------------+--------------------------+-----------+----------+-------------
 uid        | bigint                   |           | plain    |
 session_id | text                     |           | extended |
 last_ts    | timestamp with time zone |           | plain    |
View definition:
 SELECT uid,
    session_id,
    max(ts) AS last_ts
   FROM ONLY stream1
  WHERE arrival_timestamp > (clock_timestamp() - '00:15:00'::interval) AND session_id IS NOT NULL
  GROUP BY uid, session_id;

pgload01g/test M #
```

But the number of `vactuples` increases now:
```
root@pgload01g ~ # pglog -f | fgrep 'vac: ' | fgrep -e recent_sessions_mrel0 -e pg_toast_16456
< 2015-09-04 11:23:33.334 MSK >DEBUG:  recent_sessions_mrel0: vac: 14146739 (threshold 2010), anl: 25450880 (threshold 18145)
< 2015-09-04 11:23:33.334 MSK >DEBUG:  pg_toast_16456: vac: 0 (threshold 200), anl: 0 (threshold 50)
< 2015-09-04 11:23:40.837 MSK >DEBUG:  recent_sessions_mrel0: vac: 14310690 (threshold 2010), anl: 25722340 (threshold 18145)
< 2015-09-04 11:23:48.564 MSK >DEBUG:  recent_sessions_mrel0: vac: 14477531 (threshold 2010), anl: 26007976 (threshold 18145)
< 2015-09-04 11:23:48.564 MSK >DEBUG:  pg_toast_16456: vac: 0 (threshold 200), anl: 0 (threshold 50)
< 2015-09-04 11:23:56.230 MSK >DEBUG:  recent_sessions_mrel0: vac: 14624064 (threshold 2010), anl: 26270248 (threshold 18145)
< 2015-09-04 11:24:04.114 MSK >DEBUG:  recent_sessions_mrel0: vac: 14791270 (threshold 2010), anl: 26544388 (threshold 18145)
< 2015-09-04 11:24:04.114 MSK >DEBUG:  pg_toast_16456: vac: 0 (threshold 200), anl: 0 (threshold 50)
^C
root@pgload01g ~ #
```

Autovacuum workers are running but still they seem to be really short-living and in the logs I see the following:
```
root@pgload01g ~ # pglog -n10050000 | fgrep -A1 'tuple concurrently updated'
< 2015-09-04 11:29:14.775 MSK >ERROR:  tuple concurrently updated
< 2015-09-04 11:29:14.775 MSK >CONTEXT:  automatic vacuum of table "test.public.recent_sessions_mrel0"
--
< 2015-09-04 11:29:30.203 MSK >ERROR:  tuple concurrently updated
< 2015-09-04 11:29:30.203 MSK >CONTEXT:  automatic vacuum of table "test.public.recent_sessions_mrel0"
--
< 2015-09-04 11:29:45.183 MSK >ERROR:  tuple concurrently updated
< 2015-09-04 11:29:45.183 MSK >CONTEXT:  automatic vacuum of table "test.public.recent_sessions_mrel0"
--
< 2015-09-04 11:30:00.054 MSK >ERROR:  tuple concurrently updated
< 2015-09-04 11:30:00.054 MSK >CONTEXT:  automatic vacuum of table "test.public.recent_sessions_mrel0"
--
< 2015-09-04 11:30:15.657 MSK >ERROR:  tuple concurrently updated
< 2015-09-04 11:30:15.657 MSK >CONTEXT:  automatic vacuum of table "test.public.recent_sessions_mrel0"
--
< 2015-09-04 11:30:30.471 MSK >ERROR:  tuple concurrently updated
< 2015-09-04 11:30:30.471 MSK >CONTEXT:  automatic vacuum of table "test.public.recent_sessions_mrel0"
--
< 2015-09-04 11:30:47.144 MSK >ERROR:  tuple concurrently updated
< 2015-09-04 11:30:47.144 MSK >CONTEXT:  automatic vacuum of table "test.public.recent_sessions_mrel0"
root@pgload01g ~ #
```

Don't you want to separate deletes of such tuples (in another bgworker, for example) and removing them with autovacuum? It seems for me to be more robust approach or am I missing something?

**Comment by derekjn** [comment]
It's expected that there will be some out-of-window tuples in the `recent_sessions_mrel0`, but it seems that the problem is now mainly due to the concurrent update issue causing the autovac to fail right away. I'll keep trying to reproduce that error; so far I haven't been able to with the `pgbench` workload you provided but it looks like we're making progress here.

**Comment by dev1ant** [comment]
Hm, it's strange that you can't reproduce the issue since I can surely reproduce it. How many TPS do you get when running pgbench? For a short-running test I get following numbers:
```
postgres@pgload01g ~ $ /usr/pgsql-9.4/bin/pgbench -n -c 4 -j 1 -T 60 -P 10 -f /tmp/insert.sql test
progress: 10.0 s, 16792.7 tps, lat 0.236 ms stddev 0.052
progress: 20.0 s, 16865.7 tps, lat 0.236 ms stddev 0.042
progress: 30.0 s, 17234.4 tps, lat 0.231 ms stddev 0.032
progress: 40.0 s, 16996.8 tps, lat 0.234 ms stddev 0.042
progress: 50.0 s, 16784.7 tps, lat 0.237 ms stddev 0.049
progress: 60.0 s, 17261.6 tps, lat 0.230 ms stddev 0.033
transaction type: Custom query
scaling factor: 1
query mode: simple
number of clients: 4
number of threads: 1
duration: 60 s
number of transactions actually processed: 1019363
latency average: 0.234 ms
latency stddev: 0.042 ms
tps = 16989.319623 (including connections establishing)
tps = 16995.486413 (excluding connections establishing)
postgres@pgload01g ~ $
```

**Comment by derekjn** [comment]
That's about the same as I'm getting. Do you see the error only if you're running the `pgbench` workload? What do the autovac logs look like if you're not doing any inserts?

**Comment by dev1ant** [comment]
No, it is not running. I've checked cron-logs and all cron-files. If it was running, there would not be such tuples:
```
pgload01g/test M # select clock_timestamp(), min(last_ts), clock_timestamp()-min(last_ts) as diff from recent_sessions_mrel0 ;
        clock_timestamp        |              min              |      diff
-------------------------------+-------------------------------+-----------------
 2015-09-04 11:22:48.434444+03 | 2015-09-04 10:56:51.902241+03 | 00:25:56.532205
(1 row)

Time: 4585.011 ms
pgload01g/test M #
```

**Comment by dev1ant** [comment]
Technically yes, but I used to insert current_timestamp in `ts` field into stream and CV takes `max(ts) as last_ts`. And right now I can see the following:
```
pgload01g/test M # select clock_timestamp(), min(_0) from recent_sessions_mrel0 ;
        clock_timestamp        |          min
-------------------------------+------------------------
 2015-09-04 18:09:36.919445+03 | 2015-09-04 13:42:22+03
(1 row)

Time: 144.207 ms
pgload01g/test M #
```

**Comment by derekjn** [comment]
Hmm, even if you vacuum?

    vacuum recent_sessions_mrel0;
    select clock_timestamp(), min(_0) from recent_sessions_mrel0 ;

**Comment by dev1ant** [comment]
And the strange thing starts here... There are not any sessions but I get error while trying to run manual vacuum:
```
pgload01g/test M # select pid, query from pg_stat_activity ;
  pid  |                     query
-------+-----------------------------------------------
  7749 | autovacuum: VACUUM public.user_activity_mrel0
 12533 | combiner2 [pldb]
 12534 | combiner1 [pldb]
 12535 | combiner0 [pldb]
 12536 | worker3 [pldb]
 12537 | worker2 [pldb]
 12538 | worker1 [pldb]
 12539 | worker0 [pldb]
 12540 | combiner3 [postgres]
 12541 | combiner2 [postgres]
 12542 | combiner1 [postgres]
 12543 | combiner0 [postgres]
 12544 | worker3 [postgres]
 12545 | worker2 [postgres]
 12546 | worker1 [postgres]
 12547 | worker0 [postgres]
 12548 | combiner3 [test]
 12549 | combiner2 [test]
 12550 | combiner1 [test]
 12551 | combiner0 [test]
 12552 | worker3 [test]
 12553 | worker2 [test]
 12554 | worker1 [test]
 12555 | worker0 [test]
 12556 | combiner3 [pipeline]
 12557 | combiner2 [pipeline]
 12558 | combiner1 [pipeline]
 12559 | combiner0 [pipeline]
 12560 | worker3 [pipeline]
 12561 | worker2 [pipeline]
 12562 | worker1 [pipeline]
 12563 | worker0 [pipeline]
  7810 | select pid, query from pg_stat_activity ;
(33 rows)

Time: 1.921 ms
pgload01g/test M # vacuum recent_sessions_mrel0;
ERROR:  XX000: tuple concurrently updated
LOCATION:  simple_heap_delete, heapam.c:3030
Time: 124.804 ms
pgload01g/test M # select pid, query from pg_stat_activity ;
  pid  |                     query
-------+-----------------------------------------------
  7749 | autovacuum: VACUUM public.user_activity_mrel0
 12533 | combiner2 [pldb]
 12534 | combiner1 [pldb]
 12535 | combiner0 [pldb]
 12536 | worker3 [pldb]
 12537 | worker2 [pldb]
 12538 | worker1 [pldb]
 12539 | worker0 [pldb]
 12540 | combiner3 [postgres]
 12541 | combiner2 [postgres]
 12542 | combiner1 [postgres]
 12543 | combiner0 [postgres]
 12544 | worker3 [postgres]
 12545 | worker2 [postgres]
 12546 | worker1 [postgres]
 12547 | worker0 [postgres]
 12548 | combiner3 [test]
 12549 | combiner2 [test]
 12550 | combiner1 [test]
 12551 | combiner0 [test]
 12552 | worker3 [test]
 12553 | worker2 [test]
 12554 | worker1 [test]
 12555 | worker0 [test]
 12556 | combiner3 [pipeline]
 12557 | combiner2 [pipeline]
 12558 | combiner1 [pipeline]
 12559 | combiner0 [pipeline]
 12560 | worker3 [pipeline]
 12561 | worker2 [pipeline]
 12562 | worker1 [pipeline]
 12563 | worker0 [pipeline]
  7810 | select pid, query from pg_stat_activity ;
(33 rows)

Time: 0.896 ms
pgload01g/test M #
```

All combiners and workers seem not to do something...

**Comment by derekjn** [comment]
Maybe it's a good idea to start with a fresh DB with only the `recent_sessions` CV to rule anything else out?

**Comment by dev1ant** [comment]
Yep, I'm going to do exactly this way.

**Comment by derekjn** [comment]
It may be a good idea to name the CV something else too, on the off chance that there is somehow a background `DELETE` operation happening.

**Comment by usmanm** [comment]
@dev1ant: We pushed out some improvements to how sliding window queries are vacuumed. I tried simulating the workload you described in this thread and everything seemed to be working alright. Whenever you get the chance, could you double check and see if the issue is resolved? Thanks!

**Comment by dev1ant** [comment]
Actually, I can't reproduce this on a fresh system with version as of fa69b5d. Right now I'm doing long-running test to verify this.

**Comment by usmanm** [comment]
Awesome! https://github.com/pipelinedb/pipelinedb/commit/0fa45a57ff7d47c8c1702a22a3dc3aa09e16cab5 should fix a few other small kinks.

**Comment by derekjn** [comment]
Thanks for all your help and responsiveness on this @dev1ant! Your insight and diagnostics made a big difference in being able to get this fairly complicated issue sorted out as quickly as possible.
